pkg/steps/new_e2e.go

```diff
@@ -0,0 +1,147 @@
+		}
+	}
+	if err := s.runSteps(ctx, s.post); err != nil {
+		errs = append(errs, fmt.Errorf("%q post steps failed: %v", s.name, err))
+	}
+	return utilerrors.NewAggregate(errs)
+}
+
+func (s *newE2ETestStep) Done() (bool, error) { return false, nil }
+func (s *newE2ETestStep) Name() string        { return s.name }
+func (s *newE2ETestStep) Description() string {
+	return fmt.Sprintf("Run cluster install %s", s.name)
+}
+
+func (s *newE2ETestStep) Requires() (ret []api.StepLink) {
+	for _, step := range append(append(s.pre, s.tests...), s.post...) {
+		ret = append(ret, api.InternalImageLink(api.PipelineImageStreamTagReference(step.Image)))
+	}
+	return
+}
+
+func (s *newE2ETestStep) Creates() []api.StepLink { return nil }
+func (s *newE2ETestStep) Provides() (api.ParameterMap, api.StepLink) {
+	return nil, nil
+}
+func (s *newE2ETestStep) SubTests() []*junit.TestCase { return nil }
+
+func (s newE2ETestStep) runSteps(ctx context.Context, steps []api.TestStep) error {
+	pods, err := s.genPods(steps)
+	if err != nil {
+		return err
+	}
+	return s.runPods(ctx, pods)
+}
+
+func (s newE2ETestStep) genPods(steps []api.TestStep) ([]coreapi.Pod, error) {
+	var ret []coreapi.Pod
+	var errs []error
+	for _, step := range steps {
+		resources, err := resourcesFor(step.Resources)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		name := s.name + "-" + step.Name
+		pod, err := generateBasePod(s.jobSpec, resources)
+		pod.ObjectMeta.Name = name
+		pod.ObjectMeta.Annotations[annotationContainersForSubTestResults] = step.Name
+		container := &pod.Spec.Containers[0]
+		container.Name = step.Name
+		container.Image = step.Image
+		container.Command = []string{"/bin/sh", "-c", "#!/bin/sh\nset -eu\n" + step.Commands}
+		container.Env = append(container.Env, []coreapi.EnvVar{
+			{Name: "NAMESPACE", Value: s.jobSpec.Namespace},
+			{Name: "JOB_NAME_HASH", Value: s.jobSpec.JobNameHash()},
+		}...)
+		if owner := s.jobSpec.Owner(); owner != nil {
+			pod.OwnerReferences = append(pod.OwnerReferences, *owner)
+		}
+		ret = append(ret, *pod)
+	}
+	if errs != nil {
+		return nil, utilerrors.NewAggregate(errs)
+	}
+	return ret, nil
+}
+
+func (s newE2ETestStep) runPods(ctx context.Context, pods []coreapi.Pod) error {
+	go func() {
+		<-ctx.Done()
+		for _, pod := range pods {
+			log.Printf("cleanup: Deleting pod %s", pod.Name)
+			if err := s.podClient.Pods(s.jobSpec.Namespace).Delete(s.name, nil); err != nil && !errors.IsNotFound(err) {
+				log.Printf("error: Could not delete %s pod: %v", pod.Name, err)
+			}
+		}
+	}()
+	for _, pod := range pods {
+		log.Printf("Executing %q", pod.Name)
+		if _, err := createOrRestartPod(s.podClient.Pods(s.jobSpec.Namespace), &pod); err != nil {
+			return fmt.Errorf("failed to create or restart %q pod: %v", pod.Name, err)
+		}
+		if err := waitForPodCompletion(s.podClient.Pods(s.jobSpec.Namespace), pod.Name, nil, false); err != nil {
+			return fmt.Errorf("%q pod %q failed: %v", s.name, pod.Name, err)
```

Comment:
Do we want, for instance, on cleanup tasks after the test has run, for all steps to try to run regardless?

Reply:
I debated whether `post` should short circuit or if the steps should `exit 0`. I went for the latter only because it made the implementation cleaner.

I believe the templates don't fail if `teardown` fails.

Reply:
It's more of a question of what API we want to expose to users here